File: binarySearch/Search81.go
func search(nums []int, target int) bool {
    n := len(nums)
    if n == 0 {
        return false
    }
    left := 0
    right := n - 1
    for left < right {
        mid := left + (right-left)/2 + 1
        if nums[mid] > nums[right] { //左半边
            //target在[left,mid)的有序区间内
            if nums[left] <= target && target < nums[mid] {
                right = mid - 1
            } else {
                left = mid
            }
        } else if nums[mid] < nums[right] {
            //target在[mid,right]
            if nums[mid] <= target && target <= nums[right] {
                left = mid
            } else {
                right = mid - 1
            }
        } else {
            //mid==right看right是不是target
            if nums[right] == target {
                return true
            }
            right--
        }
    }
    return nums[left] == target
}

File: binarySearch/Search81.py
class Solution:
    def search(self, nums: List[int], target: int) -> bool:
        left, right = 0, len(nums)-1
        res = 0
        # 尾部去重，方便确定target和mid所在区间
        while right >= 0 and nums[right] == nums[0]:
            right -= 1
        while left <= right:
            mid = left + (right-left)//2
            v = nums[mid] if (nums[mid] < nums[0]) == (target < nums[0]) else float("inf") if nums[mid] < nums[0] else float("-inf")
            if v <= target:
                res = mid
                left = mid + 1
            else:
                right = mid - 1
        return True if nums[res] == target else False